node/collation-generation/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+			Ok(Signal(ActiveLeaves(ActiveLeavesUpdate { activated, .. }))) => {
+				// follow the procedure from the guide
+				if let Some(ref config) = self.config {
+					if let Err(err) = handle_new_activations(config, &activated, ctx).await {
+						log::warn!(target: "collation_generation", "failed to handle new activations: {:?}", err);
+						return true;
+					};
+				}
+				false
+			}
+			Ok(Signal(Conclude)) => true,
+			Ok(Communication {
+				msg: CollationGenerationMessage::Initialize(config),
+			}) => {
+				// REVIEW: what happens if someone sends two initializaiton messages:
+				// panic, replace, ignore, abort?
+				// for now, we won't panic, but we'll break the run loop
+				if self.config.is_some() {
+					log::warn!(target: "collation_generation", "double initialization");
+					true
+				} else {
+					self.config = Some(config);
+					false
+				}
+			}
+			Ok(Signal(BlockFinalized(_))) => false,
+			Err(err) => {
+				log::error!(target: "collation_generation", "error receiving message from subsystem context: {:?}", err);
+				true
+			}
+		}
+	}
+}
+
+impl<Context> Subsystem<Context> for CollationGenerationSubsystem
+where
+	Context: SubsystemContext<Message = CollationGenerationMessage>,
+{
+	fn start(self, ctx: Context) -> SpawnedSubsystem {
+		let subsystem = CollationGenerationSubsystem { config: None };
+
+		let future = Box::pin(subsystem.run(ctx));
+
+		SpawnedSubsystem {
+			name: "CollationGenerationSubsystem",
+			future,
+		}
+	}
+}
+
+#[derive(Debug, derive_more::From)]
+enum Error {
+	#[from]
+	Subsystem(SubsystemError),
+	#[from]
+	OneshotRecv(oneshot::Canceled),
+	#[from]
+	Runtime(RuntimeApiError),
+	#[from]
+	Util(util::Error),
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+async fn handle_new_activations<Context: SubsystemContext>(
+	config: &CollationGenerationConfig,
+	activated: &[Hash],
+	ctx: &mut Context,
+) -> Result<()> {
+	// follow the procedure from the guide:
+	// https://w3f.github.io/parachain-implementers-guide/node/collators/collation-generation.html
+
+	for relay_parent in activated.iter().copied() {
+		let global_validation_data = request_global_validation_data_ctx(relay_parent, ctx)
+			.await?
+			.await??;
+
+		let availability_cores = request_availability_cores_ctx(relay_parent, ctx)
+			.await?
+			.await??;
+
+		// REVIEW: this section is largely cribbed from the provisioner. Is this a safe assumption to make?
+		// If yes: TODO: we should abstract away the duplication between these sections
+		for core in availability_cores {
+			let (scheduled_core, assumption) = match core {
+				CoreState::Scheduled(scheduled_core) => {
+					(scheduled_core, OccupiedCoreAssumption::Free)
+				}
+				CoreState::Occupied(occupied_core) => {
+					if let Some(scheduled_core) = occupied_core.next_up_on_available {
+						(scheduled_core, OccupiedCoreAssumption::Included)
+					} else {
+						// REVIEW: the guide doesn't mention it, but should we also handle the timed out case?
```

Comment:
I suggest we file the whole `Occupied` case for later. The exact logic that we want to employ here is going to need design of its own, because we won't want collators to wait for bitfields as a heuristic as to whether they should create a new block or not. The reason for that is that we want backing validators to have as much time to validate as possible.

AFAIK the `CandidateBacking` subsystem doesn't handle the non-`Scheduled` case anyway